### PR TITLE
GH Actions for publishing to docker hub

### PR DESCRIPTION
**Pull request by tphalp**
This PR adds a GitHub Action workflow that auto builds and publishes the docker images to Dockerhub. 

_Note, most of the commits to this PR were due to the difficult nature of debugging GH actions. I'm not aware of a way to test these offline, and as such, I had to commit changes to see if they worked. I will squash and merge so as not to muddy the history with my tomfoolery._

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -13,7 +13,7 @@ Demo="true" #skips userid check to allow access
 
 DefaultProvider="" # mx | sophtron
 
-Port="8080" #the port where the server runs
+PORT="8080" #the port where the server runs
```
Comment (tphalp) [inline .env.example:16]: Do to conventions with GutHub, any ENV var used in GH actions needs to follow standards of being UPPERCASE. That is why some of the env vars have been updated. I propose we change them all in the near future.

File: .github/workflows/docker-publish.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
+name: 'Publish Docker image'
```
Comment (tphalp) [inline .github/workflows/docker-publish.yml:1]: New docker publish workflow.

File: .github/workflows/e2e-tests.yml [hunk 3]
```diff
@@ -4,34 +4,39 @@ name: E2E Tests
 on: pull_request
 
 jobs:
+  setup-env:
```
Comment (tphalp) [inline .github/workflows/e2e-tests.yml:7]: Update e2e workflow to use new env var workflow.

File: .github/workflows/e2e-tests.yml [hunk 4]
```diff
@@ -4,34 +4,39 @@ name: E2E Tests
...
+
     steps:
       - name: Checkout
         uses: actions/checkout@v4
       
-      - uses: actions/setup-node@v3
+      - uses: actions/setup-node@v4
         with:
-          node-version: "lts/*"
+          node-version: '20'
```
Comment (tphalp) [inline .github/workflows/e2e-tests.yml:22]: As I was running this, I noticed a warning about an old version of node, so I changed this to specify the node version.

File: .github/workflows/setup-env.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+name: Setup env file contents
```
Comment (tphalp) [inline .github/workflows/setup-env.yml:1]: New workflow file that sets sup vars for env file. This allows less duplication of code.

File: .github/workflows/unit-tests.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
+name: Unit/Integration tests
```
Comment (tphalp) [inline .github/workflows/unit-tests.yml:1]: This is just a file rename for better clarity. Also added env-var setup workflow.

File: Dockerfile [hunk 7]
```diff
@@ -2,7 +2,7 @@
 # Please run `docker compose up` from the root of the project to run the docker environment for this project.
 
 FROM alpine:3.19.1
-ENV NODE_VERSION 20.11.1
+ENV NODE_VERSION 20.12.2
```
Comment (tphalp) [inline Dockerfile:5]: This update should fix one vulnerability that docker scout found.

File: README.md [hunk 8]
```diff
@@ -1,14 +1,13 @@
 # Universal Connect Widget (Application)
```
Comment (tphalp) [inline README.md:1]: Some simple typo fixes, and added more info about Docker publishing, as well as some additional details to how to run docker, and potential issues that might arise.

File: docker-compose.yaml [hunk 9]
```diff
@@ -3,44 +3,49 @@ version: '3.8'
 services:
 
   server:
-    depends_on: [cache, ui]
+    depends_on:
+      cache:
+        condition: service_healthy
+      ui:
+        condition: service_healthy
+
```
Comment (tphalp) [inline docker-compose.yaml:11]: Added additional specifics that should help fix potential race conditions that could affect us if the services take some additional time to spin up.

File: ui/Dockerfile [hunk 10]
```diff
@@ -14,7 +15,7 @@ COPY ./ ./
 RUN npm run build
 
 FROM alpine:3.19.1 as web
-ENV NODE_VERSION 20.11.1
+ENV NODE_VERSION 20.12.2
```
Comment (tphalp) [inline ui/Dockerfile:18]: Change, as referenced above, that will fix a recently found vulnerability.

File: ui/README.md [hunk 11]
```diff
@@ -1,7 +1,7 @@
 # Universal Connect Widget embedded project
 
 ## Docker
-This project is hosted on Dockerhub. It is mean to be run via docker-compose. Please follow the steps in the 
+This project is hosted on Dockerhub. It is meant to be run via docker-compose. Please follow the steps in the
```
Comment (tphalp) [inline ui/README.md:4]: Grammar fix

File: .github/workflows/e2e-tests.yml (hunk 4 above)
Comment (wesrisenmay-mx) [reply .github/workflows/e2e-tests.yml:22]: I'd prefer to keep this on lts. I'd rather have to fix things when lts gets updated, then get way behind by hardcoding a version

File: .github/workflows/e2e-tests.yml [hunk 12]
```diff
@@ -4,34 +4,39 @@ name: E2E Tests
...
-          echo WEBHOOKHOSTURL=${{ vars.WEBHOOKHOSTURL }} >> .env
-          echo LOGLEVEL=${{ vars.LOGLEVEL }} >> .env
-          echo PORT=${{ vars.PORT }} >> .env
-          echo RESOURCEPREFIX=${{ vars.RESOURCEPREFIX }} >> .env
+          
+          # Vars
+          echo -e "${{ needs.setup-env.outputs.env_vars }}" >> .env
           echo RESOURCEVERSION="" >> .env
+
+          # Secrets (can't load these from another job, due to GH security features)
```
Comment (wesrisenmay-mx) [inline .github/workflows/e2e-tests.yml:33]: I gave up on trying to share the env code after realizing how difficult it was to share secrets, but this looks like a great compromise. At least we can share the env variables

File: .github/workflows/docker-publish.yml [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    needs: [setup-env]
+
+    steps:
+      - name: 'Check out the repo'
+        uses: actions/checkout@v4
+
+      - name: 'Set up Node'
+        uses: actions/setup-node@v4
+        with:
+          node-version: '20'
```
[the remark this answers is not in the document]
Comment (tphalp) [reply .github/workflows/docker-publish.yml]: Fixed

File: .github/workflows/unit-tests.yml [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    needs: [setup-env]
+
+    steps:
+      - name: Checkout under $GITHUB_WORKSPACE
+        uses: actions/checkout@v4
+
+      - name: Set up NodeJS
+        uses: actions/setup-node@v4
+        with:
+          node-version: 20
```
[the remark this answers is not in the document]
Comment (tphalp) [reply .github/workflows/unit-tests.yml]: Fixed

**Comment by tphalp** [comment]
@wesrisenmay-mx I think I addressed all your concerns. Thanks for the feedback.

**Review by wesrisenmay-mx — APPROVED** [review]

**Pull request merged**